### DISCUSSION TITLE
Lesson 7: How to mock v0.8 chainlink contracts?

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#1260] [Q&A]
edit: I corrected some mistakes. After finally finishing Lesson 7 I have a better understanding of mocks.

# How to mock v0.8 chainlink contracts and interfaces?
I have adapted lesson 7s lottery to use v0.8 versions of the Aggregator, LinkToken, and VRF interfaces. It compiles fine and when using it with remix I am getting the right responses. I would like to implement unit tests it with python now and I am having a hard time understanding what mocks I should use or if there aren't any.

**Question:** Do I need to mock Interfaces at all? They don't do much besides forcing me to implement certain functions.

## Here are the imports from my lottery contract

```rb
import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import "@chainlink/contracts/src/v0.8/interfaces/LinkTokenInterface.sol";
import "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";
```

Looking into my `.brownie` folder I will find the following files:
`C:\Users\n4n0b1t3\.brownie\packages\smartcontractkit\chainlink-brownie-contracts@0.4.0\contracts\src\v0.8\mocks`

Distilled output
```rb
v0.8>tree /F
C:.

├───interfaces
│       AggregatorInterface.sol
│       AggregatorV2V3Interface.sol
│       AggregatorV3Interface.sol
│       AggregatorValidatorInterface.sol
│       LinkTokenInterface.sol
│       VRFCoordinatorV2Interface.sol
├───mocks
│       MockAggregator.sol
│       MockAggregatorValidator.sol
│       VRFCoordinatorMock.sol
│       VRFCoordinatorV2Mock.sol
│
├───tests
│   │   ChainlinkClientTestHelper.sol
│   │   ChainlinkTestHelper.sol
│   │   VRFConsumer.sol
│   │   VRFConsumerExternalSubOwnerExample.sol
│   │   VRFConsumerV2.sol
│   │   VRFCoordinatorV2TestHelper.sol
│   │   VRFExternalSubOwnerExample.sol
│   │   VRFMaliciousConsumerV2.sol
│   │   VRFOwnerlessConsumerExample.sol
│   │   VRFRequestIDBaseTestHelper.sol
│   │   VRFSingleConsumerExample.sol
│   │   VRFTestHelper.sol
│   │   VRFV2RevertingExample.sol
```

The mocks here seem not to do much and are very different from the mocks we are using from chainlink-mix
https://github.com/smartcontractkit/chainlink-mix/tree/main/contracts/test

## some digging
If you take a look at the chainlink-mix example code you can see that v0.8 is implemented as well:
`https://github.com/smartcontractkit/chainlink-mix/blob/main/contracts/VRFConsumerV2.sol`

## Is building my own mocks worth the trouble?
So I have three options here:

1. The smart way: give up on the v0.8 version and focus on finishing the course by doing what Patrick does.
2. The hard way: Understanding how mocks are built and making my own mocks. Learn a lot lose a lot of time.
3. The easy way: Find someone who already did this and implement the solution.

As you might guess, I am in favor of option 3, can you help me with this?

**Comment by cromewar** [comment]
You are awesome!!

**Reply by n4n0b1t3** [reply]
Thank you @cromewar! Any ideas on how to do that?

**Reply by cromewar** [reply]
Difficult to answer as I don't see a problem with using mocks for version lesser than 0.8, as they should have retro compatibility. At the end the mocks are just for testing purposes and not everything must be by force on sol 0.8. Anyways this tend to happen on new technology such as solidity (there are a lot more work to be done).
I strongly suggest just use the versions Patrick uses to learn the fundamentals and the face this problems at real development.

**Reply by cromewar** [reply]
Also this could be helpful: https://github.com/gnosis/mock-contract

**Comment by natnaelab** [comment]
Have you come up with a solution to this?

**Reply by n4n0b1t3** [reply]
unfortunately not, I just implemented everything in solidity ^0.6.6, because it's already a challenge to get through the whole course of 16 hours and I was kind of eager to get to lesson 11. There will be a Javascript oriented updated course, if I understand Patricks messages correctly. So I am waiting for this one and hope this will be on solidity >=0.8. Btw. I learned, that it is actually good to practice developing something on not the latest version. It's a skill very useful in corporate environments.

**Comment by JohnEMagee** [comment]
Due to the mac m1 issues, I just gave up at the 7ish hour mark.  There were too many links to other contracts that I couldn't make work (it was using chainlink for the random numbers issue)
